File: jsonrpcws.js
/*
jsonrpcws.js
JSON-RPC over Websocket client implementation
@author: Martin Boros <brainrape@chaosmedia.hu>

usage:

//TODO

*/

//TODO: fix inconsistencies with py
//TODO: convert to class
//TODO: doc
//TODO: licence

function JSONRPCWSService(url, service_def) {
    if(!window.console){window.console={log:function(){}}}
    if(!window.WebSocket){throw "jsonrpcws error: WebSockets not available"}
    if(!window.JSON){throw "jsonrpcws error: JSON (de)serializer not available"}

    //make `this` available in closures
    var that = this
    
    //apply the service definition
    this.local = service_def['local']
    this.onopen = service_def['onopen']
    this.onclose = service_def['onclose']
    
    //set up instance variables
    this._CALLBACK_QUEUE_SIZE = 32
    this._id = null
    this._next_id = 1
    this._callbacks = []

    // Set up WebSocket and event handlers
    this.ws = new WebSocket(url);

    this.ws.onopen = function(){
        console.log("Websocket at "+url+" connection open.")
        if(that['onopen']){ that.onopen.apply(service) }
    }

    this.ws.onmessage = function(evt) {
        //TODO: error checking
        data = JSON.parse(evt.data);
        console.log("Websocket at "+url+" got data: ", evt.data);
        if (data['method']){
            service_method = that.local[data['method']]
            that._id = data['id']
            ret = service_method.apply(that, data['params'])
            service._id = null
            if(data['id']){
                that.respond(data['id'], ret[0], ret[1])
            }
        } else if (data['result']){
            for(var i=0; i<that._callbacks.length; i++){
                if(that._callbacks[i][0] == data['id']){
                    var callback = that._callbacks.splice(i,1)[0][1]
                    callback.apply(that, [data['result'], data['error']])
                    break;
                }
            }
        } else {
            console.log("Websocket at "+url+" JSONRPC error.");
            that.close()
        }
    };

    this.ws.onerror = function(){
        console.log("Websocket at "+url+" error.");
        if(that['onclose']){ that.onclose.apply(that, error) }
    }

    this.ws.onclose = function(){
        console.log("Websocket at "+url+" connection closed.")
        if(that['onclose']){ that.onclose.apply(that, undefined) }
    }
}


JSONRPCWSService.prototype = {
    request: function(method, params, callback){
        /// Send a request
        //set id
        var id = null
        if(callback){ id = this._next_id; this._next_id += 1; }
        
        //save callback
        if(this._callbacks.length > this._CALLBACK_QUEUE_SIZE){
            this._callbacks.pop()
        }
        if(callback){
            this._callbacks = [[id, callback]].concat(this._callbacks)
        }

        //send message
        message = JSON.stringify({id:id, method:method, params:params})
        console.log("Websocket sending data: "+message)
        this.ws.send(message)
    },
    notify: function(method, params){
        /// Send a notification
        this.request (method, params, undefined)
    },
    respond: function(id, result, error){
        /// Send a response
        message = JSON.stringify({id:id, result:result, error:error})
        console.log("Websocket sending data: "+message)
        this.ws.send(message)
    },
    close: function(){
        /// Close the connection
        this.ws.close()
    }
}


    

File: jsonrpcws.py
#!/usr/bin/env python
#
# jsonrpcws.py
# 
# JSON-RPC 1.0 over Websocket server implementation
# @author: Martin Boros <brainrape@chaosmedia.hu>
#
# usage:
#
# import eventlet
# @eventlet.websocket.WebSocketWSGI
# class MyService(jsonrpcws.JsonRpcWebSocket):
#     def hello(self, name):
#         print "got hello from", name
#         if self._id: # if this is a request, return a (response, error) tuple
#             return ( ["hi, "+name], None)


#TODO: error handling
#TODO: doc
#TODO: licence

from eventlet import websocket
import simplejson


class JsonRpcError(Exception):
    pass

class JSONRPCWSService(object):
    _CALLBACK_QUEUE_SIZE = 32
    
    def __init__(self, ws):
        # Websocket handler.
        self.ws = ws
        self._id = None
        self._next_id = 1
        self._callbacks = []
        
        self.onopen()
        try:
            while 1:
                ws_message = ws.wait()
                if ws_message is None: break;
                try:
                    message = simplejson.loads(ws_message)
                except Exception:
                    raise JsonRpcError("error decoding json")
                if message.get('method',None):
                    self._onrequest(message.get("id",None),
                                    message["method"],
                                    message.get("params",[]))
                elif message.get('result'):
                    self._onresponse(message["id"],
                                     message["result"],
                                     message.get("error",None))
                else:
                    raise JsonRpcError("malformed message received")
        except JsonRpcError, e:
            self.onclose(e)
        else:
            self.onclose(None)

    def request(self, method, params, callback=None):
        '''Send JSON-RPC request. Callback will run when response arrives.'''
        id = None
        if callback:
            id = self._next_id
            self._next_id += 1
        message = simplejson.dumps({'id':id, 'method':method, 'params':params})
        self._ws.send( message )
        
        if callback:
            if len(self._callbacks) > _CALLBACK_QUEUE_SIZE:
                self._callbacks.pop()
            self._callbacks.insert(0,(id,callback))

    def notify(self, method, params):
        '''Send JSON-RPC notification'''
        message = simplejson.dumps({'id':None, 'method':method, 'params':params})
        self.ws.send( message )

    def respond(self, id, result, error=None):
        '''Send JSON-RPC response'''
        message = simplejson.dumps({'id':id, 'result':result, 'error':error})
        self.ws.send(message)
        
    def onopen():
        '''Socket open event handler.'''
        return

    def onclose(error):
        '''Socket close event handler.'''
        return

    def _onrequest(self, id, method, params):
        # Request event handler. Invokes `method` on the class instance
        if method[0] == '_': raise JsonRpcError("method names cannot start with _")
        # Get the method, bound to the service instance
        service_method = getattr(self.local, method).__get__(self)
        self._id = id
        ret = service_method(*params)
        self._id = None
        if id:
            result, error = ret
            self.respond(id, result, error)
        
    def _onresponse(self, id, result, error):
        # Response event handler.
        for c in self._callbacks:
            if c[0]==id:
                c[1](result, error)
                self._callbacks.remove(c)
